tests/test_tensor/test_optimizers.py

```diff
@@ -34,6 +35,55 @@
         not found_torch, reason='pytorch not installed'))
 
 
+@pytest.fixture
+def tagged_qaoa_tn():
+    """
+    make qaoa tensor network, with RZZ and RX tagged on a per-round basis 
+    so that these tags can be used as shared_tags to TNOptimizer
+    """
+
+    n = 8
+    depth = 4
+    terms = [ (i,(i+1)%n) for i in range(n) ]
+    gammas = qu.randn(depth)
+    betas = qu.randn(depth)
+
+    # make circuit
+    circuit_opts = {'gate_opts':{'contract':False}}
+    circ = qtn.Circuit(n, **circuit_opts)
+
+    # layer of hadamards to get into plus state
+    for i in range(n):
+        circ.apply_gate('H', i, gate_round=0)
+
+    for d in range(depth):
+        for (i, j) in terms:
+            circ.apply_gate('RZZ', -gammas[d], i, j, gate_round=d, parametrize=True)
+
+        for i in range(n):
+            circ.apply_gate('RX', betas[d] * 2, i, gate_round=d, parametrize=True)
+
+    # tag circuit for shared_tags
+    tn_tagged = circ.psi.copy()
+    variable_finder = re.compile(r'ROUND_(\d+)')
+
+    for t in tn_tagged:
+        if ('RZZ' in t.tags):
+            for tag in t.tags:
+                match = variable_finder.match(tag)
+                if match is not None:
+                    i = int(match.groups(1)[0])
+                    t.add_tag('p'+f'{2*i}')
+        elif ('RX' in t.tags):
+            for tag in t.tags:
+                match = variable_finder.match(tag)
+                if match is not None:
+                    i = int(match.groups(1)[0])
+                    t.add_tag('p'+f'{2*i+1}')
```

Comment:
Ah, this is perfect!

Reply:
Is there a way to tag individual tensors when building a circuit? I tried putting a `tags=` in the call to `apply_gate` but that didn't work because the function inside `apply_gate` was already separately being passed a `tags` argument